controllers/account.go

```diff
@@ -0,0 +1,185 @@
+// Copyright 2020 The casbin Authors. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package controllers
+
+import (
+	"fmt"
+
+	"github.com/casbin/casbin-dashboard/object"
+	"github.com/casbin/casbin-dashboard/util"
+
+	"context"
+	"encoding/json"
+	"io/ioutil"
+	"net/http"
+	"sync"
+
+	"github.com/astaxie/beego"
+	"golang.org/x/oauth2"
+)
+
+var githubEndpoint = oauth2.Endpoint{
+	AuthURL:  "https://github.com/login/oauth/authorize",
+	TokenURL: "https://github.com/login/oauth/access_token",
+}
+
+var githubOauthConfig = &oauth2.Config{
+	ClientID:     beego.AppConfig.String("GithubAuthClientID"),
+	ClientSecret: beego.AppConfig.String("GithubAuthClientSecret"),
+	RedirectURL:  "",
+	Scopes:       []string{"user:email", "read:user"},
+	Endpoint:     githubEndpoint,
+}
+
+func (c *ApiController) AuthGithub() {
+	code := c.Input().Get("code")
+	state := c.Input().Get("state")
+	addition := c.Input().Get("addition")
+	RedirectURL := c.Input().Get("redirect_url")
+	var resp Response
+	var res authResponse
+	res.IsAuthenticated = true
+
+	if state != beego.AppConfig.String("GithubAuthState") {
+		res.IsAuthenticated = false
+		resp = Response{Status: "fail", Msg: "unauthorized", Data: res}
+		c.ServeJSON()
+		return
+	}
+
+	githubOauthConfig.RedirectURL = RedirectURL
+	fmt.Print(RedirectURL, "\n")
+
+	// https://github.com/golang/oauth2/issues/123#issuecomment-103715338
+	// ctx := context.WithValue(oauth2.NoContext, oauth2.HTTPClient, httpClient)
+	token, err := githubOauthConfig.Exchange(context.Background(), code)
+	if err != nil {
+		res.IsAuthenticated = false
+		panic(err)
```

Comment:
Please avoid using `panic` function in coding, maybe you can log error here and interpret to user end in a friendly way.